mlir/lib/Dialect/Affine/Analysis/Utils.cpp

```diff
@@ -218,13 +236,17 @@ bool MemRefDependenceGraph::init() {
     unsigned n = memrefAndList.second.size();
     for (unsigned i = 0; i < n; ++i) {
       unsigned srcId = memrefAndList.second[i];
-      bool srcHasStore =
-          getNode(srcId)->getStoreOpCount(memrefAndList.first) > 0;
+      Node *srcNode = getNode(srcId);
+      bool srcMayHaveMemEffOrStore =
+          !srcNode->memEffects.empty() ||
+          srcNode->getStoreOpCount(memrefAndList.first) > 0;
       for (unsigned j = i + 1; j < n; ++j) {
         unsigned dstId = memrefAndList.second[j];
-        bool dstHasStore =
-            getNode(dstId)->getStoreOpCount(memrefAndList.first) > 0;
-        if (srcHasStore || dstHasStore)
+        Node *dstNode = getNode(dstId);
+        bool dstMayHaveMemEffOrStore =
+            !dstNode->memEffects.empty() ||
+            dstNode->getStoreOpCount(memrefAndList.first) > 0;
+        if (srcMayHaveMemEffOrStore || dstMayHaveMemEffOrStore)
```

Comment:
The check on the destination cannot be just for writes. Otherwise, it will miss out on handling cases where srcNode has load op and dstNode has loadOp and a memoryEffectOp (which has unknown side-effects). 

The following sibling loops below will merge, whereas they shouldn't, if we remove the `!dstNode->memEffects.empty()` check:
```
func.func @should_not_fuse_loops_with_side_effects(%arg0:memref<10xf32>) {
    affine.for %i = 0 to 10 {
      %0 = affine.load %arg0[%i] : memref<10xf32>
    }
    affine.for %j = 0 to 10 {
      %0 = affine.load %arg0[%j] : memref<10xf32>
     "foo"():()->()
    }
    return
  }
```

Reply:
Also changed variable names to a better representation `dstMayHaveMemEffOrStore`